### PR TITLE
[DDP] Move gradient setting to None to python

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94258
* #94320

Some workloads are seeing not all grads set to None after backward
despite grad.reset() call in reducer, moving to the python callback alleviates
the issue.

Differential Revision: [D43071397](https://our.internmc.facebook.com/intern/diff/D43071397/)

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py
Comment (wanchaol) [inline torch/testing/_internal/distributed/distributed_test.py:4612]: why we need to change the test iteration?

**Review by wanchaol — APPROVED** [review]
stamp to unblock, please fix the lint before landing

**Comment by awgu** [comment]
Do we know why calling `grad.reset()` in the reducer is not correctly setting gradients to `None`?

File: torch/testing/_internal/distributed/distributed_test.py
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py:4612]: I initially added 100 to ensure determinism, but feel like 100 iterations increases TTS a bit too much. Can revert this back if needed.

**Pull request closed without merging**
